### PR TITLE
[WIP][inductor] better heuristics found by coordesc

### PR DESCRIPTION
**Pull request by shunting314**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #97855
* #98524

I'll accumulate the perf numbers here: https://docs.google.com/spreadsheets/d/1Ga5_twazRahmS5M0g8CzCapZfB4cepy-265ffYrIjPE/edit#gid=0 

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @soumith @anijain2305 @desertfire

### CODE REVIEW

**Review by jansel — APPROVED** [review]

File: torch/_inductor/triton_heuristics.py [hunk 1]
```diff
@@ -514,7 +509,17 @@ def pointwise(size_hints, meta, tile_hint=None, filename=None):
     bs = max(256, min(numel // 128, 1024))
 
     if len(size_hints) == 1:
-        return cached_autotune([triton_config(size_hints, bs)], meta=meta)
+        configs = [triton_config(size_hints, bs)]
+        if config.max_autotune:
+            configs.extend([
+                # improve 1.832x for https://gist.github.com/shunting314/69b5055193148ade349ac7e58c85d2d9
```
Comment (ngimel) [inline torch/_inductor/triton_heuristics.py:515]: this case is probably spilling a lot with default config, we should check spills when we compile and recompile with more warps, even when max_autotune is not on.

File: torch/_inductor/utils.py [hunk 2]
```diff
@@ -912,7 +912,10 @@ def report_category(category, profile_events):
         for ev in profile_events:
             total_time += ev.self_cuda_time_ms
             percent = f"{ev.self_cuda_time_ms / wall_time_ms * 100:.2f}%"
-            rows.append([ev.key[:120], ev.self_cuda_time_ms, ev.count, percent])
+            key = ev.key
+            if key.startswith("triton_poi_fused_"):
```
Comment (ngimel) [inline torch/_inductor/utils.py:916]: why this change?

File: torch/_inductor/triton_heuristics.py [hunk 3]
```diff
@@ -582,6 +587,18 @@ def reduction(size_hints, reduction_hint=False, meta=None, filename=None):
...
                 triton_config_reduction(size_hints, 64, 64),
                 triton_config_reduction(size_hints, 8, 512),
+                # improve 1.121x for https://gist.github.com/shunting314/6267da87c6524dab29a3e33f14ff91db
+                Config({"XBLOCK": 1, "RBLOCK": 4096}, num_warps=8, num_stages=1),
+                # improve 1.074x for https://gist.github.com/shunting314/bb36b37a3d049a29ff6bada9dbe18fb8
+                Config({"XBLOCK": 2, "RBLOCK": 2048}, num_warps=8, num_stages=1),
+                # improve 1.143x for https://gist.github.com/shunting314/ae6a0d3c63409bed7759ee7bdcde01a8
+                Config({"XBLOCK": 1, "RBLOCK": 1024}, num_warps=16, num_stages=1),
+                # improve 1.033x for https://gist.github.com/shunting314/d1036afd5210e19da3144c24f3b475d8
+                Config({"XBLOCK": 1, "RBLOCK": 1024}, num_warps=8, num_stages=1),
```
Comment (ngimel) [inline torch/_inductor/triton_heuristics.py:597]: here again we can probably rely on spills information to not check configs with insufficient num_warps?

**Review by ngimel — APPROVED** [review]
I'm not against landing as is, however we should come up with ways to filter configs after compilation (based on spills) without full benchmarking

File: torch/_inductor/utils.py (hunk 2 above)
Comment (shunting314) [reply torch/_inductor/utils.py:916]: Some kernel name is too long to fit the screen. Remove the common/uninteresting prefix.

File: torch/_inductor/triton_heuristics.py (hunk 1 above)
Comment (shunting314) [reply torch/_inductor/triton_heuristics.py:515]: yep, I also checked the kernel's triton code. It does a lot of things. The default config simply based on numel can pick a too large block size.

I'll followup with a separate PR to consider number of spills

**Comment by shunting314** [comment]
> I'm not against landing as is, however we should come up with ways to filter configs after compilation (based on spills) without full benchmarking

So far in this PR I just add the promising configs found by coordinate descent tuning when applied to the top expensive pointwise and reduction kernels of a model/graph. I'll add more as I tried on more models.

The purpose is to see what overall perf improvement we can get by adding more configs (disregarding compiling time at the beginning). This is important since we don't want to improve individual kernel perf a lot but dost not improve the model training time overall 

I'll try to use other information (like number of spills you mentioned) to reduce the number of configs we should benchmark.

**Comment by ngimel** [comment]
You can get perf on this PR using ciflow/inductor-perf-compare label, but I don't think it sets max-autotuning

**Comment by shunting314** [comment]
> You can get perf on this PR using ciflow/inductor-perf-compare label, but I don't think it sets max-autotuning

Thanks for mentioning this. I didn't know we have such a label.

Does it run with and without the PR and compare the perf?
